### ISSUE TITLE
fypp with pypy throwing a FyppFatalError/NameError

### ISSUE THREAD

**Issue by MuellerSeb** [aradi/fypp#32]
When using fypp with pypy, I get the following strange error:
```
common.fypp:66: error: exception occurred when setting variable(s) 'VERSION90' to 'defined('VERSION90')' [FyppFatalError]
  error: name 'defined' is not defined [NameError]
```
([file from Fortran stdlib](https://github.com/fortran-lang/stdlib/blob/df1e2f0ed0cbe2fbd9c1f20dcb5bd1a4bba95bb2/src/common.fypp#L88))

Don't know what goes wrong here...

**Comment by MuellerSeb** [comment]
Some more context:

This happens on a conda-forge feedstock but only for `aarch64` and `ppc64le` platforms with this docker: `quay.io/condaforge/linux-anvil-cos7-x86_64`

See here for reference: https://github.com/conda-forge/mhm-feedstock/pull/15

**Comment by zerothi** [comment]
I think the smallest reproducer is:
```fortran
program t
print *, ${NAME}$
end program
```
Trying to fyppifying this with:
```shell
$> fypp -DNAME=hello test.fypp
error: exception at evaluating 'hello' in definition for 'NAME' [FyppFatalError]
error: name 'hello' is not defined [NameError]
```

The problem is that `fypp` tries to resolve the name (if it ain't stringified).
This is a bit problematic when one, e.g. wishes to use `iso_fortran_env` names for variables:
`-DINT_TYPE=int32`, and a fortran code: `integer(${INT_TYPE}$)` (this won't work!).
I can see why this is hard to figure out, is the name a variable, or a string?

The solution is something obscure like this:
```shell
$> fypp -DNAME="'hello'" test.fypp
```
to fully escape the name as a string (either single string will be shell-resolved and not passed to `fypp`).

I don't know of an elegant solution other than automatically stringifying anything it doesn't name-resolve, but even then it is problematic because you don't know whether it *should* be a name, or whether it really should be a string.

**Comment by aradi** [comment]
Yes, Fypp currently evaluates everything after the `=` sign as a Python expression. So, yes, you have to pass the string-delimiters either with `-DNAME="'hello'"` or with `-DNAME=\"hello\"`. We could easily introduce a second kind of define (like in CMake), so that the right hand side is automatically interpreted as a string `-SNAME=hello` and once could enforce conversion (if needed) with the explicit naming of the conversion function `-SNAME:bool=True`.

Would that be something which would make your life easier?

**Comment by zerothi** [comment]
> Yes, Fypp currently evaluates everything after the `=` sign as a Python expression. So, yes, you have to pass the string-delimiters either with `-DNAME="'hello'"` or with `-DNAME=\"hello\"`. We could easily introduce a second kind of define (like in CMake), so that the right hand side is automatically interpreted as a string `-SNAME=hello` and one could even enforce conversion (if needed) with the explicit naming of the conversion function, like `-SNAME:bool=True`.
> 
> Would that be something which would make your life easier?

Oh, yes, I actually thought of this, thinking I shouldn't propose it !!! ;) haha...

I think this could be nice, in particular the delimiters required for string handling in cmake projects are annoying, but not only that, it is also error prone, say if the user them-selves put in quotation that gets caught by cmake, what should one do then?

My thought on initial variable names would be:

- `:INT`
- `:FLOAT`
- `:PATH` (necessary?)
- `:STRING` (or just `:STR`)
- `:BOOL`
- `:NONE` for a define only variable? Or is it overkill? Sometimes it can be hard to know whether a definition actually has a meaning somewhere...
- `:ARRAY(<any of the above>[, delim=,])`, say `-DH:ARRAY(str,;)=int32;int64` or `-DH:ARRAY(str)=int32,int64` or similar?

Might have omitted some? Or are some just overkill?

**Comment by aradi** [comment]
That's definitely an overkill, and brought me to change my mind about the converters. :laughing: 

My suggestion would be to implement the `-S` option first without any conversion possibility. This would already allow to remedy the problems with the tricky quotation when passing strings. For all other the cases (e.g. where a conversion is needed), one can use the good old  `-D` option with any valid Python expression.

As an example, given the file `test.fypp` with
```
$:array
```
The command
```
fypp -m numpy -Darray='numpy.array([1, 2, 3], numpy.float32)' test.fypp
```
results in
```
[1. 2. 3.]
```
Note, the quotation in the command above is only needed to prevent bash to valute the square brackets...
So, probably, we should not try to add complicated converters for `-S`, if the same functionality can be easily used with the `-D` option already. Let's just fix the only issue, which is more complicated then it needs to be: the passing of string values when defining entitites on the command line.

**Comment by zerothi** [comment]
Agreed, let's keep it simple, the `fypp -SVAR=hello` seems like a nice and simple addition!

**Comment by aradi** [comment]
@zerothi  After thinking about that a while (and trying to find the time for it 😉), I came up with a proposal in #52, which is similar to the one discussed above but with slightly different logic:

* It introduces the option `--define-mode`. When set to `expr`, the `-D` options work as always (with expression evaluation). When set to `str`, the values of the `-D` options are treated as string literals, so no quoting is needed.
 
* Additionally, the option `-S` (set) is introduced, which always does expression evaluation independent of the `--define-mode` setting.

The rationale is that many build systems by default use the `-D` option automatically when passing definitions to pre-processors, so the user should be able to use `-D` both ways, depending what is better suited for the given scenario. Also, on the long term, one could turn the default of `--define-mode` to `str`, which would have been probably a more reasonable choice from the beginning. And in order to still maintain the possibility to set evaluated expressions independently of `--define-mode`, the `-S` option with value evaluation is introduced. (So, `-D` becomes flexible, and `-S` takes the role of the old `-D`.)

Any comments are welcome. I'll leave the PR open and not merge for a few weeks, as I also have to rethink once more, whether this is the best strategy or whether we should rather use `-S` for string literal defines and `-D` for evaluated defines, as discussed above.
